lib/OpenQA/WebAPI/Controller/API/V1/Search.pm

```diff
@@ -123,27 +110,55 @@ sub query {
             }
             $last_filename = $filename;
             push(@results, {occurrence => "$basename/$filename", contents => $contents});
+            # For the purposes of the limit, all lines in the same file count as one
+            $cap--;
+            last if $cap < 1;
         }
-        $cap -= scalar @results;
-        last if $cap < 1;
+    }
+    return \@results;
+}
 
-        # Job templates
-        my $last_group = undef;
-        my $like       = {like => "%${keywords}%"};
-        my $templates  = $self->schema->resultset('JobTemplates')
-          ->search({-or => {name => $like, description => $like}}, {limit => $cap});
-        while (my $template = $templates->next) {
-            my $contents = $template->name . "\n" . $template->description;
-            if ($template->group->id == $last_group) {
-                $results[-1]->{contents} .= "\n$contents";
-                next;
-            }
-            $last_group = $template->group->id;
-            push(@results, {occurrence => $template->group->name, contents => $contents});
+sub _search_job_templates {
+    my ($self, $keywords, $limit) = @_;
+
+    my @results;
+    my $last_group = -1;
+    my $like       = {like => "%${keywords}%"};
+    my $templates  = $self->schema->resultset('JobTemplates')
+      ->search({-or => {name => $like, description => $like}}, {limit => $limit});
+    while (my $template = $templates->next) {
+        my $contents = $template->name . "\n" . $template->description;
+        if ($template->group->id == $last_group) {
+            $results[-1]->{contents} .= "\n$contents";
+            next;
         }
-        $cap -= scalar @results;
-        last if $cap < 1;
+        $last_group = $template->group->id;
+        push(@results, {occurrence => $template->group->name, contents => $contents});
     }
+    return \@results;
+}
+
+sub query {
+    my ($self) = @_;
+
+    my $validation = $self->validation;
+    $validation->required('q');
+    return $self->reply->validation_error({format => 'json'}) if $validation->has_error;
+
+    # Allow n queries per minute, per user (if logged in)
+    my $lockname = 'webui_query_rate_limit';
+    if (my $user = $self->current_user) { $lockname .= $user }
+    return $self->render(json => {error => 'Rate limit exceeded'}, status => 400)
+      unless $self->app->minion->lock($lockname, 60, {limit => $self->app->config->{'rate_limits'}->{'search'}});
+
+    my $cap = $self->app->config->{'global'}->{'search_results_limit'};
+    my @results;
+    my $keywords = $validation->param('q');
+
+    my $perl_module_results = $self->_search_perl_modules($keywords, $cap);
+    $cap -= scalar @{$perl_module_results};
+    push @results, @{$perl_module_results};
+    push @results, @{$self->_search_job_templates($keywords, $cap)};
```

Comment:
do we need to add a check such as `if $cap > 0` when calling `_search_job_templates`?

Reply:
I'm relying on the actual query here to enforce the limit